0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,180 @@
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
+
+Third we dealt with the problem that there is not always an `address_hint` (i.e. Bob does not need to know the `address_hint` of Alice).
+We added the `address_hint` extension table that relates to `swap` on the `counter_party_peer_id` field (rather than on the `swap_local_id`).
+
+Note that the role could be factored out of the table, but we decided against it because it adds unnecessary complexity for now.
+
+What remains is a simple table for `swap` which contains only the information which all swaps have in common: `role`, `alpha_protocol`, `beta_protocol` and `counter_party_peer_id`.
+
+=== Protocols
+
+[[HanProtocols]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/han_swap_protocols.puml&fmt=svg[HanProtocols]
+
+In order to store details for each swap such as data we need for creating the HTLCs for each protocol we introduce `\{protocol\}_swap_details`, where `\{protocol\}` can be `han_bitcoin`, ``han_ethereum`, `herc20` or `halight` for now.
+Similar to the `Swaps` table, we add information which is only known at a later point of time into a sub-table.
+This resulted in quite an excess of sub-tables such as `han_bitcoin/ethereum_redeem_identity`, `han_bitcoin/ethereum_refund_identity`, etc.
+
+Some notable sub-tables:
+
+* `han_bitcoin/ethereum_swap_detail` contains data needed to create the HTLCs. We decided to give this table a generated primary key, rather than using `swap_local_id` as the primary key to allow same chain swaps.
+* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
+
+One could optimize here and combine all ledger specific identities into an own table `han_bitcoin_identity` to hold all Bitcoin `refund` and `redeem` identities.
+This would mean though that we would need an additional unique id in those sub-tables, i.e. we can't use the `fk` as `pk`.
+
+* `han_bitcoin/ethereum_ledger_state`: is a new table which we introduced in this database schema.
+It records observed ledger events such as `fund`/`redeem`/`refund`-transaction.
+We extracted the states into `han_state` in order to reduce duplication in each table and not have an extra column for the event-type (`fund`/`redeem`/`refund`).
+This allows us to reuse these states among all `han_*` protocols.
```

Comment:
I would argue against this, see here: https://github.com/comit-network/comit-rs/issues/2533

Reply:
Can you elaborate on this please? I don't see the relation to the linked ticket.

Reply:
The ticket suggests that there should be no abstract "han" concept which means no reuse between those protocols, at least not on a conceptual level. On an implementation detail level, we can decide to reuse logic that is the same but I don't think there is much to reuse apart from names :D

Reply:
I thought the definition of `han_*` protocols was that they all share the same 4 actions (fund/redeem/refund).

Reply:
In my head it is like @bonomat says as well. Do we really want to get rid of the `HAN` abstraction ("protocol family") completely?

Reply:
> I thought the definition of `han_*` protocols was that they all share the same 4 actions (fund/redeem/refund).

Yes that was the idea and I think you can still think of them like that if you want.
More interestingly though, it seems like that additional layer of abstraction is not worth it in many ways:

- In Rust code, it just leads to a large number of generics
- In the API, you always you need to look for the combination of `han` and `ethereum`/`bitcoin` anyway, otherwise you don't know how to semantically interpret the data.
- Same goes for the database.

It feels like this abstract protocol (iiee pfui, object-orientation) doesn't give us much because we hardly ever _abstract_ over it.
Interestingly, even though the actions are the same, the client doesn't really care about this because those actions are just "executable".

In other words: It doesn't feel like these protocols actually have something in common apart from using HTLCs (but that applies to all `H*` protocols). There is nothing special about the Ether HTLC that it **shares** with the Bitcoin one but differentiates it from the one for ERC20.

Doing an atomic swap on Bitcoin follows a completely different protocol than the Ethereum one does. We just invented shared terminology but that is about it :)

Conceptually, I am liking more and more the idea of cnd supporting a flat hierarchy of protocols where some are compatible with each other.

HErc20 and HAlight are the norm, HAN is the exception and should be aligned :)

Reply:
I've split up `han_swap_details` to `han_bitcoin` and `han_ethereum`